### ISSUE TITLE
Port twisted.python.deprecate to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5916]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5916|
|Type|enhancement|
|Created|2012-08-24 19:09:38Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-python3-5916-3|

This is needed by `SynchronousTestCase` to implement the `assertWarns` functionality.  Specifically, trial uses:

  * deprecated
  * getDeprecationWarningString
  * warnAboutFunction
  * deprecatedModuleAttribute

However, the last two are only used to warn about deprecated functionality, so we can probably skip them and remove that deprecated functionality instead.

<details><summary>Searchable metadata</summary>

```
trac-id__5916 5916
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_deprecate_python3_5916_3 branches/deprecate-python3-5916-3
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1345835378000000 1345835378000000
changetime__1346425912000000 1346425912000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [35432]) Branching to 'deprecate-python3-5916-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [35433]) Port some parts of `twisted.python.deprecate` that trial really wants

refs [#5916](https://github.com/twisted/twisted/issues/5916)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Moved tests for `deprecated` and `getVersionString` and `getDeprecationWarningString` into `twisted/python/test/test_deprecatepy3.py`
  1. Changed some strings in those tests which depend on the module in which the code is defined.
  1. Changed one test using `assertWarns` to use `warnings.catch_warnings` instead.
  1. Moved implementation of those methods into `twisted/python/_deprecatepy3.py`
  1. Also moved `_fullyQualifiedName` and changed its method handling around a little to work with Python 3.3 changes (should really do this in a separate branch, but waiting for other reflect changes to land first).

Tested locally, but didn't force a build yet, buildbot is too busy.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [35442]) Filename can be pyc.
refs [#5916](https://github.com/twisted/twisted/issues/5916)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [35443]) Branching to 'deprecate-python3-5916-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [35445]) Move the fullyQualifiedName tests into test_reflectpy3; adjust them to work in that position.  Fix a couple more Python 3 incompatibilities in that function.

refs [#5916](https://github.com/twisted/twisted/issues/5916)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Okay, since [#5914](https://github.com/twisted/twisted/issues/5914) landed, I did the `fullyQualifiedName` tests properly.  In addition to the list of things I mentioned above, the branch now moves the `fullyQualifiedName` tests into `test_reflectpy3.py`, adjusts them to use available names (it imports a lot of random extra names; the specific names are not relevant, just the kind of objects they refer to), and fixes another couple bugs in the `fullyQualifiedName` port.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-python3-5916-3)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [35447]) Either side can be a pyc

refs [#5916](https://github.com/twisted/twisted/issues/5916)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst set owner to @exarkun|
|-|-|

  1. The test that compares `reflect` to `"twisted.python._reflectpy3"` and says it's a trick is bad. Either compare `_reflectpy3` (you can reimport it) or just do compat.
  1. I don't understand the line at top of test_reflectpy3 module about how the function should be imported from `reflect`.
  1. Add `__future__` imports.

Please fix the above, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [35457]) Improve the comment about the _fullyQualifiedName import.  Also remove the tricky test, which is redundant anyway.

refs [#5916](https://github.com/twisted/twisted/issues/5916)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [35458]) Add future imports

refs [#5916](https://github.com/twisted/twisted/issues/5916)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [35459]) Merge deprecate-python3-5916-3

Author: exarkun
Reviewer: itamar
Fixes: [#5916](https://github.com/twisted/twisted/issues/5916)

Port parts of `twisted.python.deprecation` to Python 3.  The parts ported are
the parts needed by trial to implement its warning-related APIs.
